File: java-version/src/week2/day1/B2667.java
package week2.day1;
//https://www.acmicpc.net/problem/2667
//단지번호 붙히기

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;

public class B2667 {
    static int map [][];
    static boolean visited[][];
    static  int countPerDanji;

    static  int dx [] ={0,0,1,-1};
    static  int dy [] ={1,-1,0,0};

    public static void dfs(int row, int col) {
        visited[row][col] = true;
        countPerDanji++;

        for (int i = 0; i < 4; i++) {
            int newX = row + dx[i];
            int newY = col + dy[i];

            if (newX >= 0 && newX < map.length && newY >= 0 && newY < map[newX].length) {
                if (map[newX][newY] == 1 && !visited[newX][newY]) {
                    dfs(newX, newY);
                }
            }
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        map = new int[n][n];
        visited = new boolean[n][n];

        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < n; j++) {
                map[i][j] = line.charAt(j) - '0';
            }
        }

        ArrayList<Integer> countList = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if(map[i][j]==1&&visited[i][j]==false){
                    countPerDanji=0;
                    dfs(i,j);
                    countList.add(countPerDanji);
                }
            }
        }

        System.out.println(countList.size());
        Collections.sort(countList);
        for (Integer i : countList) {
            System.out.println(i);
        }

    }
}

File: codingTest/2week-assignment/day1/B2667.py
def dfs(row, col):
    global countPerDanji
    visited[row][col] = True
    countPerDanji += 1

    # Directions: up, down, right, left
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]

    for i in range(4):
        newX = row + dx[i]
        newY = col + dy[i]

        if 0 <= newX < n and 0 <= newY < n:
            if map[newX][newY] == 1 and not visited[newX][newY]:
                dfs(newX, newY)

n = int(input())
map = []
visited = [[False for _ in range(n)] for _ in range(n)]

for _ in range(n):
    row_num = input()
    row_value = [int(i) for i in row_num]
    map.append(row_value)

countList = []

for i in range(n):
    for j in range(n):
        if map[i][j] == 1 and not visited[i][j]:
            countPerDanji = 0
            dfs(i, j)
            countList.append(countPerDanji)


print(len(countList))
for count in sorted(countList):
    print(count)